File: src/main/java/toker/panel/aop/interceptor/DefaultStringValueInterceptor.java
package toker.panel.aop.interceptor;

import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Pointcut;
import toker.panel.annotation.DefaultStringValue;

import java.lang.reflect.Method;

@Aspect
public class DefaultStringValueInterceptor {

    @Pointcut("execution(public String *(..)) && @annotation(toker.panel.annotation.DefaultStringValue)")
    public void getting() { }

    public String aroundGetting(ProceedingJoinPoint joinPoint) throws Throwable {
        String value = (String) joinPoint.proceed();
        if (value == null) {
            Class<?> type = joinPoint.getSignature().getDeclaringType();
            try {
                Method getter = type.getDeclaredMethod(joinPoint.getSignature().getName());
                return getter.getAnnotation(DefaultStringValue.class).value();
            } catch (NoSuchMethodException ignored) { }
        }
        return value;
    }
}

File: src/main/java/toker/panel/aop/interceptor/DefaultStringValueInterceptor.kt
package toker.panel.aop.interceptor

import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Pointcut
import toker.panel.annotation.DefaultStringValue

@Aspect
class DefaultStringValueInterceptor {
    @Pointcut("execution(public String *(..)) && @annotation(toker.panel.annotation.DefaultStringValue)")
    fun getting() {
    }

    @Throws(Throwable::class)
    fun aroundGetting(joinPoint: ProceedingJoinPoint): String? {
        val value = joinPoint.proceed() as String?
        if (value == null) {
            val type = joinPoint.signature.declaringType
            try {
                val getter = type.getDeclaredMethod(joinPoint.signature.name)
                return getter.getAnnotation(DefaultStringValue::class.java).value
            } catch (ignored: NoSuchMethodException) {
            }
        }
        return value
    }
}